scene/gui/text_edit.cpp

```diff
@@ -2156,408 +2204,577 @@ void TextEdit::_new_line(bool p_split_current_line, bool p_above) {
 	}
 
 	begin_complex_operation();
-
-	bool first_line = false;
-	if (!p_split_current_line) {
-		deselect();
-		if (p_above) {
-			if (caret.line > 0) {
-				set_caret_line(caret.line - 1, false);
-				set_caret_column(text[caret.line].length());
+	Vector<int> caret_edit_order = get_caret_index_edit_order();
+	for (const int &i : caret_edit_order) {
+		bool first_line = false;
+		if (!p_split_current_line) {
+			deselect(i);
+			if (p_above) {
+				if (get_caret_line(i) > 0) {
+					set_caret_line(get_caret_line(i) - 1, false, true, 0, i);
+					set_caret_column(text[get_caret_line(i)].length(), i == 0, i);
+				} else {
+					set_caret_column(0, i == 0, i);
+					first_line = true;
+				}
 			} else {
-				set_caret_column(0);
-				first_line = true;
+				set_caret_column(text[get_caret_line(i)].length(), i == 0, i);
 			}
-		} else {
-			set_caret_column(text[caret.line].length());
 		}
-	}
 
-	insert_text_at_caret("\n");
+		insert_text_at_caret("\n", i);
 
-	if (first_line) {
-		set_caret_line(0);
+		if (first_line) {
+			set_caret_line(0, i == 0, true, 0, i);
+		}
 	}
-
 	end_complex_operation();
 }
 
 void TextEdit::_move_caret_left(bool p_select, bool p_move_by_word) {
-	// Handle selection
-	if (p_select) {
-		_pre_shift_selection();
-	} else if (selection.active && !p_move_by_word) {
-		// If a selection is active, move caret to start of selection
-		set_caret_line(selection.from_line);
-		set_caret_column(selection.from_column);
-		deselect();
-		return;
-	} else {
-		deselect();
-	}
-
-	if (p_move_by_word) {
-		int cc = caret.column;
-		// If the caret is at the start of the line, and not on the first line, move it up to the end of the previous line.
-		if (cc == 0 && caret.line > 0) {
-			set_caret_line(caret.line - 1);
-			set_caret_column(text[caret.line].length());
+	for (int i = 0; i < carets.size(); i++) {
+		// Handle selection
```

Comment:
```suggestion
		// Handle selection.
```
Could as well fix the comments if they are being moved.